# ***Purpose :*** count copy number deletion and insertion allele frequence from VCFs

***Input :*** Structure variantions in VCF format generated by lumpy and sv-typer



**Output :**

* CNV information for each sample (no overlapped regions)
* CNV allele frequecy for a group of samples (no overlapped regions)



**Commands :**

> \# convert vcf file to tsv format and split overlapped regions and update the genotype information for each region block
>
>  vcf2tsv -r -f CHROM,POS,END,GT,SAMPLE  sample.sv.vcf.gz | CNV_regions_format_per_sample -min 1000 > sample_min1K_CNV.bed

> \# merge all the formated bed files and calculate the CNV















